src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+    #[doc = " to the additional level of indirection."]
+    #[doc = ""]
+    #[doc = " @param callback Interrupt handler function."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub unsafe fn subscribe(&self, callback: impl FnMut(u32) + Send + 'static) -> Result<(), EspError> {
+        enable_isr_service()?;
+
+        //self.unsubscribe();
+        let callback: Box<dyn FnMut(u32) + 'static> = Box::new(callback);
+        ISR_HANDLERS[self.unit as usize] = Some(callback);
+        esp!(esp_idf_sys::pcnt_isr_handler_add(
+            self.unit,
+            Some(Self::handle_isr),
+            self.unit as *mut esp_idf_sys::c_types::c_void,
+        ))?;
+        Ok(())
+    }
+
+    #[doc = " @brief Remove ISR handler for specified unit."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn unsubscribe(&self) -> Result<(), EspError> {
+        unsafe {
+            esp!(esp_idf_sys::pcnt_isr_handler_remove(self.unit))?;
+            ISR_HANDLERS[self.unit as usize] = None;
+        }
+        Ok(())
+    }
+
+    unsafe extern "C" fn handle_isr(data: *mut esp_idf_sys::c_types::c_void) {
+        let unit = data as pcnt_unit_t;
+        if let Some(f) = &mut ISR_HANDLERS[unit as usize] {
+            let mut value = 0u32;
+            esp!(esp_idf_sys::pcnt_get_event_status(
+                unit,
+                &mut value as *mut u32
+            )).expect("failed to fetch event status!");
+            f(value);
+        }
+    }
+}
+
+impl Drop for Pcnt {
+    fn drop(&mut self) {
+        let _ = self.counter_pause();
+        let _ = self.intr_disable();
+        unsafe {ISR_HANDLERS[self.unit as usize] = None};
+        unit_deallocate(self.unit)
+    }
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static ISR_SERVICE_ENABLED: core::sync::atomic::AtomicBool =
+    core::sync::atomic::AtomicBool::new(false);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static PCNT_CS: crate::task::CriticalSection = crate::task::CriticalSection::new();
+
+fn enable_isr_service() -> Result<(), EspError> {
+    use core::sync::atomic::Ordering;
+
+    if !ISR_SERVICE_ENABLED.load(Ordering::SeqCst) {
+        let _ = PCNT_CS.enter();
+
+        if !ISR_SERVICE_ENABLED.load(Ordering::SeqCst) {
+            esp!(unsafe { esp_idf_sys::pcnt_isr_service_install(0) })?;
+
+            ISR_SERVICE_ENABLED.store(true, Ordering::SeqCst);
+        }
+    }
+
+    Ok(())
+}
+
+static mut ISR_HANDLERS: [Option<Box<dyn FnMut(u32)>>; esp_idf_sys::pcnt_unit_t_PCNT_UNIT_MAX as usize] = [
+    None, None, None, None, 
+    #[cfg(not(esp32s3))]
+    None,
+    #[cfg(not(esp32s3))]
+    None,
+    #[cfg(not(esp32s3))]
+    None,
+    #[cfg(not(esp32s3))]
+    None, 
+];
+
+static mut PCNT_UNITS: [Option<UnitHandle>; esp_idf_sys::pcnt_unit_t_PCNT_UNIT_MAX as usize] = [
```

Comment:
These seem to be the peripherals. Please examine the code of the other drivers, and align your code with how other peripherals declare and implement peripherals.

Reply:
Implemented Pcnt trait and peripherals.